### ISSUE TITLE
Add Support for header tokens

### ISSUE THREAD

**Issue by pepie** [luanpotter/audioplayers#129]

**Comment by pepie** [comment]
It would be nice to be able to pass an auth token, for secured remote content.
is this possible?

**Comment by erickzanardo** [comment]
I am not sure, would to do some research. I wonder if this is something achieved on the native platform

**Comment by zencubic** [comment]
Would like to have this feature too!
@erickzanardo On android ist not too complicated (https://stackoverflow.com/a/29128037, https://developer.android.com/reference/android/media/MediaPlayer.html#setDataSource(android.content.Context,%2520android.net.Uri,%2520java.util.Map%3Cjava.lang.String,%2520java.lang.String%3E)) but I don't know about iOS.
It would probably be best if you pass your own HTTP-Header in dart code

**Comment by zencubic** [comment]
Maybe this is helpful for iOS: https://stackoverflow.com/questions/15456130/add-custom-header-field-in-request-of-avplayer

**Comment by s98113161** [comment]
I need this feature too!

**Comment by MasjZam** [comment]
I've implemented the Android side - anyone have experience with Objective C willing to do the iOS side for this?

**Comment by mihael998** [comment]
Somebody has some news for this? I would need it too.

**Comment by alectogeek** [comment]
Please add this feature!

**Comment by qianciling** [comment]
I need this feature too!

**Comment by WinXaito** [comment]
+1

**Comment by BartusZak** [comment]
HTTPS headers for web please :D

**Comment by AhemadAbbasVagh** [comment]
any updates on HTTPS headers for web?

**Comment by keyur005** [comment]
Pl add this feature!

**Comment by synchronisator** [comment]
I try to play a file from a Nextcloud. 
To do this, i need to add a Basic Authentication Header.
This is not possible at the moment.
Because in web it is not possible to save the file to disk and play it from there, and playing UInt8List is also not implemented, it is at this moment not possible to play a file from nextcloud in web.

Please add this feature.
Thanks

**Comment by Gustl22** [comment]
Proposal: implement #374 by downloading the file beforehand and then can use header tokens to get the file.

**Comment by cnkygmr** [comment]
any update ?

**Comment by maximeburri** [comment]
+1
Any update ?

**Comment by 1250422131** [comment]
> Proposal: implement #374 by downloading the file beforehand and then can use header tokens to get the file, so we don't rely on the platform implementation.

Downloading files in advance is reasonable and effective, but we need to be aware of one issue: most of the time, we prefer online audio to be cached in segments rather than caching the entire file before playback. This allows us to save some data usage and enables faster playback of online music.

**Comment by jvaleski** [comment]
+1

**Comment by jvaleski** [comment]
audioplayers uses AVPlayerItem under the hood for iOS, and Apple has gone way out of their way to prevent arbitrary (auth for example) header additions to the AVURLAsset's that AVPlayerItem uses. as best I can tell, you'd have to subclass AVURLAsset, creating your own custom URL scheme, and implement your own delegation of the resource loading to manage all the network interactions yourself.

feeling like all roads lead to accessing remote URLs that require custom headers via an HTTP lib, saving data locally, then having audioplayers play the local file. the alternative is a pretty big overhaul of how audioplayers currently implements UrlSource in order to get custom headers in there.

**Comment by tom-barone** [comment]
One solution could be adding another `Source` implementation that downloads the file to the device as part of the `setOnPlay` method. A variant of this approach has been working for me.

A quick and dirty example:

```dart
class DownloadedFileSource extends Source {
   /// The URL to download the file from
  final String url;
  /// The filename to save the file as
  final String filename;
  /// Optional headers to send with the request
  final Map<String, String>? headers;

  @override
  Future<void> setOnPlayer(AudioPlayer player) async {
    // Example: Using the `path_provider` package
    final Directory cacheDir = await getTemporaryDirectory();
    final String filePath = '${cacheDir.path}/$filename';
    var client = http.Client();
    try {
      var request = http.Request('GET', Uri.parse(url));
      if (headers != null) {
        request.headers.addAll(headers!);
      }
      var response = await client.send(request);
      if (response.statusCode == 200) {
        var file = File(filePath);
        await response.stream.pipe(file.openWrite());
      } else {
        throw Exception(
          'Failed to download file with status: ${response.statusCode}',
        );
      }
    } finally {
      client.close();
    }
    return player.setSourceDeviceFile(filePath, mimeType: mimeType);
  }
//...
}
```

Happy to open a PR, but would be keen to hear a maintainers opinion on caching first. 
Should the library take care of caching / disposing of the downloaded file under the hood?

**Comment by Gustl22** [comment]
Ideally it should be done on the platform side within the native players to properly support buffering and live streams:
See as an example the Flutter Video Player: https://github.com/flutter/packages/blob/8c287e99d62396402f9a8eb89d9704a5ba6cb8c0/packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart#L150

**Comment by tom-barone** [comment]
Makes sense @Gustl22. I'll leave the above as just a workaround for now.

**Comment by cuongmedia** [comment]
This is easy to do, why does it take so long?
